enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,568 @@
+// +listType=map
+// +listMapKey=name
+// +optional
+PreCreate []LifecycleHook `json:"preCreate,omitempty"`
+~~~
+
+Creation of the resource will be blocked until the `preCreate` hook is removed from the `machine.machine.openshift.io` instance.
+
+### Workflow Description
+
+#### Installation
+1. OpenShift administrator reserves IP addresses for installation.
+2. OpenShift administrator constructs `install-config.yaml` to define an nmstate configuration for each node that will receive a static IP address.
+3. OpenShift administrator initiates an installation with `openshift-install create cluster`.  
+4. The installer will proceed to:
+- provision bootstrap and control plane nodes with specified IP configuration
+- create machine resources containing specified IP configuration
+5. Once the machine API controllers become active, the compute machine resources will be rendered with the specified IP configuration.
+
+#### Scaling new Nodes without `machinesets`
+1. OpenShift administrator reserves IP addresses for new nodes to be scaled up.
+2. OpenShift administrator constructs machine resource to define an nmstate configuration for each new node that will receive a static IP address.
+3. OpenShift administrator initiates the creation of new machines by running `oc create -f machine.yaml`.  
+4. The machine API will render the nodes with the specified IP configuration.
+
+#### Scaling new Nodes with `machinesets`
+1. OpenShift administrator configures a machineset with a `preCreate` lifecycle hook.
+
+Example of a `machineset` configured to configure the `preCreate` lifecycle hook on machines it creates.
+~~~yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+metadata:
+  name: static-machineset-worker
+  namespace: openshift-machine-api
+  labels:
+    machine.openshift.io/cluster-api-cluster: cluster
+spec:
+  replicas: 0
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-cluster: cluster
+      machine.openshift.io/cluster-api-machineset: static-machineset-worker
+  template:
+    metadata:
+      labels:
+        machine.openshift.io/cluster-api-cluster: cluster
+        machine.openshift.io/cluster-api-machine-role: worker
+        machine.openshift.io/cluster-api-machine-type: worker
+        machine.openshift.io/cluster-api-machineset: static-machineset-worker
+    spec:
+      lifecycleHooks:
+        preCreate:
+          - name: ipamController
+            owner: network-admin
+~~~
+
+2. OpenShift administrator or machine autoscaler scales `n` machines
+3. Controller watches machine resources created with the a `preCreate` lifecycle hook which matches
+the expected name/owner.
+4. Controller updates machine providerSpec with IP configuration
+5. Controller sets `preTerminate` lifecycle hook
+6. Controller removes `preCreate` lifecycle hook
+
+On scale down, the controller will recognize a machine is being deleted and check for a `preTerminate`
```

Comment:
just a thought, if we end up changing `preCreate` to `preProvisioning` perhaps we want to make this closer to the machine phases as well, eg `preDeleting` or something?

Reply:
`preTerminate` already exists and was chosen because there are two stages to deleting. There's the drain, and then the instance terminate. A pre-deleting would not allow us to distinguish between those two stages